src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp

```diff
@@ -723,6 +729,15 @@ std::shared_ptr<Repeated> Snapshot::tryMergeTriangles(const std::vector<Group::G
         return {};
     }
 
+    if (prods.size() < m_ctx.keep_blocks) {
+        // In some cases (specifically mixed precision) we could have merged
+        // a small number of huge groups which consumed other legit repeated blocks
+        // due to having a different repeated tag due to unique weights precision combination
+        // from the rest of the model.
+        // This check was added to prevent that and shouldn't affect other models.
```

Comment:
I don't get how the comment matches the check itself. You refer to some different repeated tag but check size against threshold. How is this supposed to work?

Reply:
Lets say we have 10 blocks with tag "aaa" and 3 blocks of "bbb". When we are trying to merge blocks further those 10 "aaa" blocks cannot be merged due to having only producers with the same "aaa". But for 3 "bbb" blocks their producers have "aaa" block -> thus the merge is possible. We merge and assign a new tag for those "ccc". Then we can merge again and again due to the tag being different from the beginning. This would consume a large amount of blocks into 3 huge blocks.

This check prevent such case in the least invasive way - in the CleanupUniques() pass we would still drop those 3 blocks due to the same check. But at that point they are already huge and consumed legit blocks. I suggest we prevent such merges here.

Reply:
Put this explanation in the comment